add a test module with some test API (#1908)
pinojs/pino: docs/api.md

* See [`logger.version`](#version)

## Test
Pino provides a test module with some useful APIs that you can use importing it as the following
```js
require('pino').test
// OR
const { test } = require('pino')
```

<a id="pino-test-sink"></a>
### `pino.test.sink({ destroyOnError = false, emitErrorEvent = false }) => Transform`

> Create a pino destination stream to easily spy on over the logs created by pino.

Reviewer: ```suggestion
Create a Pino destination stream to easily inspect the logs processed by Pino.
```

> Create a stream using `split2` so that each line is a chunk and each chunk is JSON parsed.

Reviewer: The purpose of this method should be explained. How the method does what it does is not useful information in the documentation.

const stream = pino.test.sink()
const logger = pino(stream)

logger.info('hello world')

stream.once('data', (data) => {
  console.log(chunk.msg) // 'hello world'
  console.log(chunk.level) // 30
})
```

Destroy the stream on error
```js
const pino = require('pino')

const stream = pino.test.sink({ destroyOnError: true })
stream.write('helloworld')
stream.end()

stream.once('close', () => {
  console.log('close event') // "close event"
})
```

Destroy and send error event on error
```js
const pino = require('pino')

const stream = pino.test.sink({ destroyOnError = false, emitErrorEvent = false })
stream.write('helloworld')
stream.end()

stream.on('error', (err) => {
  console.log(err) // Unexpected token h in JSON at position 0
})

stream.on('close', () => {
  console.log('close event') // "close event"
})
```

Send error event on error
```js
const pino = require('pino')

const stream = pino.test.sink({ emitErrorEvent = false })
stream.write('helloworld')
stream.end()

stream.on('error', (err) => {
  console.log(err) // Unexpected token h in JSON at position 0
})
```

<a id="pino-test-once"></a>
### `pino.test.once(stream, expected, is) => Promise<void>`

> Assert that a single pino log is expected.

Reviewer: ```suggestion
Assert that a single log is expected.
```

> Assert chunk is expected in the stream using the `stream.once` event listener. 

Reviewer: What is "chunk"? What is the purpose of this method?

```js
const pino = require('pino')

const stream = pino.test.sink()
const instance = pino(stream)

instance.info('hello world')

const expected = { msg: 'hello world', level: 30 }
await pino.test.once(stream, expected) // doesn't throw a diff error
await pino.test.once(stream, { msg: 'bye world', level: 30 }) // throw a diff error

// OR logging an object
instance.info({ hello: 'world', hi: 'world' })
await pino.test.once(stream, { hello: 'world', hi: 'world', level: 30 }) // doesn't throw a diff error
await pino.test.once(stream, { hello: 'world', level: 30 }) // throw a diff error

// OR using a custom assert function
function is (received, expected, msg) {
  if (received.msg !== expected.msg) {
    throw new Error(`expected msg ${expected.msg} doesn't match the received one ${received.msg}`)
  }
}

await pino.test.once(stream, expected, is) // doesn't throw an error
await pino.test.once(stream, { msg: 'bye world', level: 30 }, is) // throw an error
```

<a id="pino-test-consecutive"></a>
### `pino.test.consecutive(stream, expected, is) => Promise<void>`

> Assert that consecutive pino's logs are expected.

Reviewer: ```suggestion
Assert that consecutive logs are expected.
```

> Assert consecutive chunks is expected in the stream using the `for await...of` loop. 

Reviewer: Again, what are "chunks"? Why would someone use this method?

```js
const pino = require('pino')

const stream = pino.test.sink()
const instance = pino(stream)

instance.info('hello world')
instance.info('hi world')

const expected = [
  { msg: 'hello world', level: 30 },
  { msg: 'hi world', level: 30 }
]

await pino.test.consecutive(stream, expected) // doesn't throw a diff error
await pino.test.consecutive(stream, [{ msg: 'bye world', level: 30 }]) // throw a diff error

// OR logging an object
instance.info({ hello: 'world' })
instance.info({ hi: 'world' })
await pino.test.consecutive(stream, [
  { hello: 'world', level: 30 },
  { hi: 'world', level: 30 }
])

// OR using a custom assert function
function is (received, expected, msg) {
  if (received.msg !== expected.msg) {
    throw new Error(`expected msg ${expected.msg} doesn't match the received one ${received.msg}`)
  }
}

await pino.test.consecutive(stream, expected, is) // doesn't throw an error
await pino.test.consecutive(stream, [{ msg: 'bye world', level: 30 }], is) // throw an error
```

> Both `pino.test.once` and `pino.test.consecutive` use the internal `check` function

Reviewer: What is "the internal `check` function"?

Author: let me remove it

> - that asserts the `chunk.time`, `chunk.pid` and `chunk.hostname`, so that you is avoided to do it

Reviewer: What is "chunk"? Also, overall, the sentence needs some rewording. What is it asserting about the specified log properties? "so that you is avoided to do it" I believe is unnecessary.

Author: yep

const instance = pino(stream)

instance.info('hello world')
instance.info('hi world')

const expected = [
  { msg: 'hello world', level: 30 },
  { msg: 'hi world', level: 30 }
]

await pino.test.consecutive(stream, expected) // doesn't throw a diff error
await pino.test.consecutive(stream, [{ msg: 'bye world', level: 30 }]) // throw a diff error

// OR logging an object
instance.info({ hello: 'world' })
instance.info({ hi: 'world' })
await pino.test.consecutive(stream, [
  { hello: 'world', level: 30 },
  { hi: 'world', level: 30 }
])

// OR using a custom assert function
function is (received, expected, msg) {
  if (received.msg !== expected.msg) {
    throw new Error(`expected msg ${expected.msg} doesn't match the received one ${received.msg}`)
  }
}

await pino.test.consecutive(stream, expected, is) // doesn't throw an error
await pino.test.consecutive(stream, [{ msg: 'bye world', level: 30 }], is) // throw an error
```

Both `pino.test.once` and `pino.test.consecutive` internally
- assert log message `time` is less than or equal to the current time
- assert log message `pid` matches the current process id
- assert log message `hostname` matches the current hostname

> - uses the default `deepStrictEqual` assert function of the `node:assert` module.

Reviewer: This part applies to two sections but is only present in one of the sections. So people are likely to miss it if they only reach the `pino.test.once` section. It should be present in both sections and specific to each.

See
* [Transform](https://nodejs.org/api/stream.html#class-streamtransform)

> * [split2](https://github.com/mcollina/split2#readme)

Reviewer: Why would someone need to reference these external resources? What is this "See" section referring to?

Author: no longer needed; let me remove it.
